### PR TITLE
bugfix: native histogram: exemplars index out of range

### PR DESCRIPTION
**Pull request by krajorama**
Fixes: #1607 
Fixes: #1605 

fix: native histogram: Simplify and fix addExemplar

mdIdx was redundant when len(exemplars)>1, so got rid of it, rIdx
is enough.

Don't compare timestamp of incoming exemplar to timestamp of
minimal distance exemplar. Most of the time the incoming exemplar
will be newer. And if not, the previous code just replaced an
exemplar one index after the minimal distance exemplar. Which had
an index out of range bug, plus is essentially random.

Contains an unoptimized fix for #1605 , see discussion / fix in #1609 

```
panic: runtime error: index out of range [10] with length 10

goroutine 10 [running]:
github.com/prometheus/client_golang/prometheus.(*nativeExemplars).addExemplar(0xc0002307c8, 0xc00055f4f0)
	/home/krajo/go/github.com/krajorama/client_golang/prometheus/histogram.go:1791 +0x1969
github.com/prometheus/client_golang/prometheus.(*histogram).updateExemplar(0xc000230700, 0x3ff22643022060a2, 0x0, 0xc00054fcb0)
	/home/krajo/go/github.com/krajorama/client_golang/prometheus/histogram.go:1140 +0x117
github.com/prometheus/client_golang/prometheus.(*histogram).ObserveWithExemplar(0xb64180?, 0xc00054fcb0?, 0xbe246f?)
	/home/krajo/go/github.com/krajorama/client_golang/prometheus/histogram.go:770 +0x6a
github.com/prometheus/client_golang/prometheus.TestNativeHistogramConcurrency.func1.2({0xc0005d0000, 0x46e4, 0x420820?})
	/home/krajo/go/github.com/krajorama/client_golang/prometheus/histogram_test.go:1058 +0x17b
created by github.com/prometheus/client_golang/prometheus.TestNativeHistogramConcurrency.func1 in goroutine 21
	/home/krajo/go/github.com/krajorama/client_golang/prometheus/histogram_test.go:1050 +0x325
```

### CODE REVIEW

**Comment by krajorama** [comment]
cc @fatsheep9146

File: prometheus/histogram.go
Comment (beorn7) [inline prometheus/histogram.go:1686]: As said in the commit description, let's discuss the optimization of this in #1609.

**Review by beorn7 — APPROVED** [review]
LGTM, but @fatsheep9146 might have an opinion here, too.

Also, leaving it to the maintainers @ArthurSens @kakkoyun @bwplotka to make the final call.

**Review by ArthurSens — APPROVED** [review]
LGTM

**Comment by bwplotka** [comment]
Does it needs a minor 1.20 release with this? Or we are ok with 1.21 in a few months or so?

**Comment by krajorama** [comment]
> Does it needs a minor 1.20 release with this? Or we are ok with 1.21 in a few months or so?

I'd love to get a minor release so we can do clean dependency update in Mimir.

**Comment by ArthurSens** [comment]
I'll keep this open for another day or so just in case Bartek or Kemal want to give another review. I create another release, but then we need to point this PR to the `release-1.20` branch

File: prometheus/histogram.go
Comment (krajorama) [reply prometheus/histogram.go]: we agreed on the solution in 1609, which is now in here.

**Review by fatsheep9146 — APPROVED** [review]
Thanks for picking this bug! 
This fix is LGTM.

File: prometheus/histogram.go
Comment (bwplotka) [inline prometheus/histogram.go:1786]: I guess this is then guarding the case when nIdx is the last element? Then again, comment 1766 is kind of wrong, or misleading.

**Review by bwplotka — COMMENTED** [review]
Some readability suggestion, some question likely due to my lack of context for the details. Plus please rebase on `release-1.20` for this to be in minor patch. Otherwise LGTM!

**Comment by krajorama** [comment]
@bwplotka I've updated the comments in the code with more explanations and my understanding of the code :)

**Comment by bwplotka** [comment]
I forced this PR to be against `release-1.20` branch.. which brought new commits from main. Do you mind rebasing your commits to `release-1.20` branch? (:

**Review by bwplotka — APPROVED** [review]
Thanks! (just rebase)

**Pull request merged**
